app/src/main/java/com/github/sdp_begreen/begreen/dialog/ContestMapDialog.kt

```diff
@@ -0,0 +1,383 @@
+ */
+class ContestMapDialog(private val listener: ContestMapDialogListener) : DialogFragment(),
+    OnMapReadyCallback {
+
+    private val contestMapDialogViewModel by viewModels<ContestMapDialogViewModel>()
+
+    private lateinit var fusedLocationClient: FusedLocationProviderClient
+    private lateinit var map: GoogleMap
+
+    private val requestPermissionLauncher =
+        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
+            if (isGranted) {
+                displayLocation()
+            } else {
+                Toast.makeText(
+                    requireActivity(),
+                    getString(R.string.location_permissions_not_granted),
+                    Toast.LENGTH_LONG
+                ).show()
+            }
+        }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.dialog_create_contest_map, container, false)
+
+        val mapFragment: SupportMapFragment =
+            childFragmentManager.findFragmentById(R.id.create_contest_map) as SupportMapFragment
+
+        mapFragment.getMapAsync(this)
+
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(requireActivity())
+
+        val locationButton = view.findViewById<MaterialButton>(R.id.create_contest_location_button)
+        val radiusButton = view.findViewById<MaterialButton>(R.id.create_contest_radius_button)
+
+        setupLocationButton(locationButton)
+        setupRadiusButton(radiusButton)
+        setupButtonBackgroundToggle(view)
+        setupApproveButton(view)
+        setupCancelButton(view)
+        return view
+    }
+
+    override fun onMapReady(map: GoogleMap) {
+        this.map = map
+        // Start by checking that the user has the permission to use his location to center the map
+        addUserLocationLayer()
+        setupAddMarker()
+        setupRedrawCircleUponMarkerChange()
+        setupDragMarkerListener()
+        setupInitMarkerAndCircle()
+    }
+
+    /**
+     * Helper function to initially draw the marker and the circle on the map
+     */
+    private fun setupInitMarkerAndCircle() {
+        var initLocation: LatLng? = null
+        var initRadius = 0.0
+        arguments?.also { arg ->
+            initLocation = arg.getParcelable<CustomLatLng?>(LOCATION_ARG)?.toMapLatLng()
+            initRadius = arg.getDouble(RADIUS_ARG)
+        }
+
+        if (initLocation != null && initRadius > 0.0) {
+            addLocationMarker(initLocation!!)
+            addRadiusMarker(addDistanceToPos(initLocation!!, initRadius))
+            drawCircle(initLocation!!, initRadius)
+        }
+    }
+
+    /**
+     * Helper function to add the user location, and request permission if not already granted
+     */
+    private fun addUserLocationLayer() {
+
+        if (ContextCompat.checkSelfPermission(
```

Comment:
I don't know is @LucaCarroz will merge his PR in the meanwhile, but if so, there is a method hasPermissions that checks the permissions.

Reply:
I will have a look at it thanks